docs/benchmarkdotnet.md

```diff
@@ -0,0 +1,332 @@
+
+
+## Main Concepts
+
+Benchmarking is really hard (especially microbenchmarking), you can easily make a mistake during performance measurements.
+BenchmarkDotNet will protect you from the common pitfalls (even for experienced developers) because it does all the dirty work for you:
+
+* it generates an isolated project per runtime (with boilerplate code)
+* it builds the project in `Release` (using Roslyn or dotnet cli)
+* it runs every benchmark in a stand-alone process (to achieve process isolation and avoid side effects)
+* it estimates the perfect invocation count per iteration (based on `IterationTime`)
+* it warms-up the code
+* it evaluates the overhead
+* it runs multiple iterations of the method until the requested level of precision is met
+* it consumes the benchmark result to avoid dead code elimination
+* it prevents from inlining of the benchmark by wrapping it with a delegate
+* it prints the results to the console in GitHub markdown, so you can just copy-paste the printed table to GitHub
+* it exports the results to `.\BenchmarkDotNet.Artifacts\results` so you can store them for later use.
+
+A few useful links for you:
+
+* If you want to know more about BenchmarkDotNet features, check out the [Overview Page](http://benchmarkdotnet.org/Overview.htm).
+* If you want to use BenchmarkDotNet for the first time, the [Getting Started](http://benchmarkdotnet.org/GettingStarted.htm) will help you.
+* If you want to ask a quick question or discuss performance topics, use the [gitter](https://gitter.im/dotnet/BenchmarkDotNet) channel.
+
+## Running the Benchmarks
+
+### Prerequisites
+
+To run the benchmarks, you need to [download dotnet cli](./prerequisites.md) first.
+
+### Interactive Mode
+
+To run the benchmarks in interactive mode you have to execute `dotnet run -f $targetFrameworkMoniker` in the folder with benchmarks project.
+
+```cmd
+C:\Projects\performance\src\benchmarks\micro> dotnet run -f netcoreapp3.0
+Available Benchmarks:
+  #0   Burgers
+  #1   ByteMark
+  #2   CscBench
+  #3   LinqBenchmarks
+  #4   SeekUnroll
+  #5   Binary_FromStream<LoginViewModel>
+  #6   Binary_FromStream<Location>
+  #7   Binary_FromStream<IndexViewModel>
+  #8   Binary_FromStream<MyEventsListerViewModel>
+  #9   Binary_FromStream<CollectionsOfPrimitives>
+  #10  Binary_ToStream<LoginViewModel>
+  ..... // the list continues
+
+You should select the target benchmark(s). Please, print a number of a benchmark (e.g. '0') or a contained benchmark caption (e.g. 'Burgers'):
+```
+
+And select one of the benchmarks from the list by either entering its number or name.
+
+### Command Line
+
+#### Filtering the Benchmarks
+
+You can filter the benchmarks using `--filter $globPattern` console line argument.
+
+The glob patterns are applied to full benchmark name: namespace.typeName.methodName. Examples:
+
+* `dotnet run -f netcoreapp3.0 --filter BenchmarksGame*` - will run all the benchmarks from BenchmarksGame namespace
```

Comment:
nit: Have the examples like this:
1. Run all the benchmarks from BenchmarksGame namespace
```cmd
dotnet run -f netcoreapp3.0 --filter BenchmarksGame*
```
2. Run all the benchmarks with type name Richards
```cmd
dotnet run -f netcoreapp3.0 --filter *.Richards.*
```
3.

Reply:
good suggestion, done